### ISSUE TITLE
It is kind of strict to require every param to be `requires_grad=True`

### ISSUE THREAD

**Issue by chenzhekl** [pytorch/pytorch#1266]
Hi,

I find that optimizer enforces every parameter to be `requires_grad=True`. However, if I used such a network where the first layer was a pre-trained `AlexNet` with every parameter set to `requires_grad = False`, and rest layers were left `requires_grad = True`, pytorch would complain `optimizing a parameter that doesn't require gradients`. Because in `model.parameters()`, some parameters do not require gradients.

Is there anything wrong with my method of freezing part of the model? or would it be better to enforce that not all parameters are set `requires_grad=False` .

**Comment by fmassa** [comment]
There has been some discussion about that, but I can't find it now (I'm on my phone).

Note that you can filter the parameters so that only the parameters that requires gradient are passed to the optimizer.
```python
filter(lambda x: return x.requires_grad, model.parameters())
```

**Comment by chenzhekl** [comment]
@fmassa Thank you for your help. That's exactly what I am doing. If this issue is a duplication, may I close it?

**Comment by fmassa** [comment]
Leave it open for the moment (as I haven't linked to the other thread).

**Comment by soumith** [comment]
this is an explicit choice to avoid silly mistakes. @apaszke took a call to keep this as is (I asked that the restriction be relaxed)

**Comment by n0obcoder** [comment]
filter(lambda x: return x.requires_grad, model.parameters()) 
is this the best way to train the parameters with requires_grad = True? what if i use
just the models.parameteres() ? do the 2 methods give the same output? i think the fuilter method wil be less time consuming biut the output will be the same? pls confirm. Thanks in advance  :D

**Comment by distant1219** [comment]
@n0obcoder I think they are the same way to train with freezing some layers. The output will be same and the filter is used to avoid silly operation.
